File: 0090-subsets-ii/0090-subsets-ii.cpp
class Solution {
public:    
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        set<vector<int>>subset;
        int n=nums.size();
        vector<int>temp;
        getAllSubset(0,n,nums,temp,subset);
        vector<vector<int>>result;
        for(auto it:subset){
            result.push_back(it);
        }
        return result;
    }
    void getAllSubset(int idx,int n,vector<int>nums,vector<int>temp,set<vector<int>>&result){
        if(idx==n){
            sort(temp.begin(),temp.end());
            result.insert(temp);
            return;
        }
        temp.push_back(nums[idx]);
        getAllSubset(idx+1,n,nums,temp,result);
        temp.pop_back();
        getAllSubset(idx+1,n,nums,temp,result);

    }
    
};

File: 0090-subsets-ii/0090-subsets-ii.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>>result=new ArrayList<>();
        Arrays.sort(nums);
        int n=nums.length;
        getAllSubsets(0,n,nums,new ArrayList<>(),result);
        return result;
    }

    private void getAllSubsets(int idx, int n, int[] nums, List<Integer> ls, List<List<Integer>> result) {
        if(idx==n){
            result.add(new ArrayList<>(ls));
            return;
        }
        
        if(ls.size()==0 || ls.get(ls.size()-1)!=nums[idx]){
            getAllSubsets(idx+1,n,nums,ls,result);
        }
        ls.add(nums[idx]);
        getAllSubsets(idx+1,n,nums,ls,result);
        ls.remove(ls.size()-1);
        
    }
}